File: src/main/java/eladkay/quaritum/common/item/chalk/ItemChalk.java
package eladkay.quaritum.common.item.chalk;

import eladkay.quaritum.api.lib.LibNBT;
import eladkay.quaritum.api.util.ItemNBTHelper;
import eladkay.quaritum.client.core.ModelHandler;
import eladkay.quaritum.common.block.ModBlocks;
import eladkay.quaritum.common.item.base.ItemMod;
import eladkay.quaritum.common.lib.LibLocations;
import eladkay.quaritum.common.lib.LibNames;
import net.minecraft.block.Block;
import net.minecraft.block.SoundType;
import net.minecraft.block.state.IBlockState;
import net.minecraft.client.renderer.color.IItemColor;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.EnumDyeColor;
import net.minecraft.item.ItemBlock;
import net.minecraft.item.ItemDye;
import net.minecraft.item.ItemStack;
import net.minecraft.util.EnumActionResult;
import net.minecraft.util.EnumFacing;
import net.minecraft.util.EnumHand;
import net.minecraft.util.SoundCategory;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;

public class ItemChalk extends ItemMod implements ModelHandler.IColorProvider, ModelHandler.ICustomLogHolder {
    private static final String[] COLORS = new String[17];

    static {
        for (EnumDyeColor dye : EnumDyeColor.values())
            COLORS[dye.ordinal()] = "chalk" + (capitalizeFirst(dye.toString())).replace("Silver", "LightGray");
        COLORS[16] = "chalkTempest";
    }

    public ItemChalk() {
        super(LibNames.CHALK, COLORS);
        addPropertyOverride(LibLocations.FLAT_CHALK, (stack, world, entityLivingBase) -> ItemNBTHelper.getBoolean(stack.copy(), LibNBT.FLAT, false) ? 1.0f : 0.0f);
        setMaxStackSize(1);
    }

    public static String capitalizeFirst(String str) {
        return str.substring(0, 1).toUpperCase() + str.substring(1);
    }

    public static EnumDyeColor byMetadata(int meta) {
        if (meta == 16) return null;
        return EnumDyeColor.byMetadata(meta);
    }

    @Override
    public EnumActionResult onItemUse(ItemStack stack, EntityPlayer player, World world, BlockPos blockPos, EnumHand hand, EnumFacing side, float hitX, float hitY, float hitZ) {
        IBlockState iblockstate = world.getBlockState(blockPos);
        Block block = iblockstate.getBlock();

        if (!block.isReplaceable(world, blockPos)) {
            blockPos = blockPos.offset(side);
        }

        if (stack.stackSize != 0 && player.canPlayerEdit(blockPos, side, stack) && world.canBlockBePlaced(ModBlocks.chalk, blockPos, false, side, null, stack)) {
            int i = stack.getMetadata();
            IBlockState iblockstate1 = ModBlocks.chalk.onBlockPlaced(world, blockPos, side, hitX, hitY, hitZ, i, player);

            if (placeBlockAt(stack, player, world, blockPos, side, hitX, hitY, hitZ, iblockstate1)) {
                SoundType soundtype = ModBlocks.chalk.getSoundType();
                world.playSound(player, blockPos, soundtype.getPlaceSound(), SoundCategory.BLOCKS, (soundtype.getVolume() + 1.0F) / 2.0F, soundtype.getPitch() * 0.8F);
            }

            return EnumActionResult.SUCCESS;
        }
        else
        {
            return EnumActionResult.FAIL;
        }
    }

    public boolean placeBlockAt(ItemStack stack, EntityPlayer player, World world, BlockPos pos, EnumFacing side, float hitX, float hitY, float hitZ, IBlockState newState) {
        if (!world.setBlockState(pos, newState, 3)) return false;

        IBlockState state = world.getBlockState(pos);
        if (state.getBlock() == ModBlocks.chalk) {
            ItemBlock.setTileEntityNBT(world, player, pos, stack);
            ModBlocks.chalk.onBlockPlacedBy(world, pos, state, player, stack);
        }

        return true;
    }

    @Override
    @SideOnly(Side.CLIENT)
    public IItemColor getColor() {
        return new IItemColor() {
            @Override
            public int getColorFromItemstack(ItemStack stack, int tintIndex) {
                return ItemNBTHelper.getBoolean(stack, LibNBT.FLAT, false) && stack.getItemDamage() < 16 ? ItemDye.DYE_COLORS[15- stack.getItemDamage()] : 0xFFFFFF;
            }
        };
    }

    @Override
    public String customLog() {
        return ModelHandler.modlength + " |  Variants by dye color";
    }

    @Override
    public String customLogVariant(int variantID, String variant) {
        return "";
    }

    @Override
    public int sortingPrecedence() {
        return 1;
    }

    @Override
    public boolean shouldLogForVariant(int variantID, String variant) {
        return false;
    }
}

File: src/main/java/eladkay/quaeritum/common/item/chalk/ItemChalk.kt
package eladkay.quaeritum.common.item.chalk

import com.teamwizardry.librarianlib.features.base.item.IItemColorProvider
import com.teamwizardry.librarianlib.features.base.item.ItemMod
import com.teamwizardry.librarianlib.features.helpers.ItemNBTHelper
import com.teamwizardry.librarianlib.features.utilities.client.TooltipHelper
import eladkay.quaeritum.api.lib.LibNBT
import eladkay.quaeritum.common.block.ModBlocks
import eladkay.quaeritum.common.lib.LibLocations
import eladkay.quaeritum.common.lib.LibNames
import eladkay.quaeritum.common.lib.arrayOfStrings
import eladkay.quaeritum.common.lib.capitalizeFirst
import net.minecraft.block.state.IBlockState
import net.minecraft.client.util.ITooltipFlag
import net.minecraft.entity.Entity
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.EnumDyeColor
import net.minecraft.item.ItemBlock
import net.minecraft.item.ItemDye
import net.minecraft.item.ItemStack
import net.minecraft.util.EnumActionResult
import net.minecraft.util.EnumFacing
import net.minecraft.util.EnumHand
import net.minecraft.util.SoundCategory
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World
import net.minecraftforge.fml.relauncher.Side
import net.minecraftforge.fml.relauncher.SideOnly

class ItemChalk : ItemMod(LibNames.CHALK, *ItemChalk.COLORS), IItemColorProvider {
    override val itemColorFunction: ((ItemStack, Int) -> Int)?
        get() = { stack, _ -> if (ItemNBTHelper.getBoolean(stack.copy(), LibNBT.FLAT, false) && stack.itemDamage < 16)
            ItemDye.DYE_COLORS[15 - stack.itemDamage] else 0xFFFFFF }

    val block = ModBlocks.chalk

    init {
        addPropertyOverride(LibLocations.FLAT_CHALK) { stack, world, entityLivingBase -> if (ItemNBTHelper.getBoolean(stack.copy(), LibNBT.FLAT, false)) 1.0f else 0.0f }
        setMaxStackSize(1)
    }

    @SideOnly(Side.CLIENT)
    override fun addInformation(stack: ItemStack, world: World?, tooltip: MutableList<String>, advanced: ITooltipFlag) {
        TooltipHelper.addToTooltip(tooltip, getUnlocalizedName(stack) + ".desc")
    }

    override fun onItemUse(player: EntityPlayer, worldIn: World, pos: BlockPos, hand: EnumHand, facing: EnumFacing, hitX: Float, hitY: Float, hitZ: Float): EnumActionResult {
        var pos = pos
        val iblockstate = worldIn.getBlockState(pos)
        val block = iblockstate.block

        if (!block.isReplaceable(worldIn, pos)) {
            pos = pos.offset(facing)
        }

        val itemstack = player.getHeldItem(hand)

        if (!itemstack.isEmpty && player.canPlayerEdit(pos, facing, itemstack) && worldIn.mayPlace(this.block, pos, false, facing, null as Entity?)) {
            val i = this.getMetadata(itemstack.metadata)
            val iblockstate1 = this.block.getStateForPlacement(worldIn, pos, facing, hitX, hitY, hitZ, i, player, hand)

            if (placeBlockAt(itemstack, player, worldIn, pos, facing, hitX, hitY, hitZ, iblockstate1)) {
                val soundtype = worldIn.getBlockState(pos).block.getSoundType(worldIn.getBlockState(pos), worldIn, pos, player)
                worldIn.playSound(player, pos, soundtype.placeSound, SoundCategory.BLOCKS, (soundtype.getVolume() + 1.0f) / 2.0f, soundtype.getPitch() * 0.8f)
            }

            return EnumActionResult.SUCCESS
        } else {
            return EnumActionResult.FAIL
        }
    }

    fun placeBlockAt(stack: ItemStack, player: EntityPlayer, world: World, pos: BlockPos, side: EnumFacing, hitX: Float, hitY: Float, hitZ: Float, newState: IBlockState): Boolean {
        if (!world.setBlockState(pos, newState, 3)) return false

        val state = world.getBlockState(pos)
        if (state.block === ModBlocks.chalk) {
            ItemBlock.setTileEntityNBT(world, player, pos, stack)
            ModBlocks.chalk.onBlockPlacedBy(world, pos, state, player, stack)
        }

        return true
    }


    companion object {
        private val COLORS: Array<String> = arrayOfStrings(EnumDyeColor.values().size)

        init {
            for (dye in EnumDyeColor.values())
                COLORS[dye.ordinal] = "chalk" + dye.toString().capitalizeFirst().replace("Silver", "LightGray")
        }


        fun byMetadata(meta: Int): EnumDyeColor? {
            if (meta == 16) return null
            return EnumDyeColor.byMetadata(meta)
        }
    }
}
